akka-http-core/src/main/scala/akka/http/scaladsl/Http.scala

```diff
@@ -658,26 +662,29 @@ object Http extends ExtensionId[HttpExt] with ExtensionIdProvider {
   final case class IncomingConnection(
     localAddress: InetSocketAddress,
     remoteAddress: InetSocketAddress,
-    flow: Flow[HttpResponse, HttpRequest, Unit]) {
+    flow: Flow[HttpResponse, HttpRequest, Promise[Option[Option[InetSocketAddress]]]]) {
 
     /**
      * Handles the connection with the given flow, which is materialized exactly once
      * and the respective materialization result returned.
      */
-    def handleWith[Mat](handler: Flow[HttpRequest, HttpResponse, Mat])(implicit fm: Materializer): Mat =
-      flow.joinMat(handler)(Keep.right).run()
+    def handleWith[Mat](handler: Flow[HttpRequest, HttpResponse, Mat])(implicit fm: Materializer): Mat = {
+      val fillMeInAndUserMat = flow.joinMat(handler)(Keep.both).run()
+      fillMeInAndUserMat._1.success(Some(Some(remoteAddress)))
+      fillMeInAndUserMat._2
+    }
```

Comment:
Agreed, I tried to do what you did locally, but I think both the attribute AND the Source.maybe solution suffers from the same problem: it is too easy to forget to set it.